src/libraries/System.Linq/src/System/Linq/Take.cs

```diff
@@ -20,6 +20,169 @@ public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> sourc
+            Index end = range.End;
+            bool isStartIndexFromEnd = start.IsFromEnd;
+            bool isEndIndexFromEnd = end.IsFromEnd;
+            int startIndex = start.Value;
+            int endIndex = end.Value;
+            Debug.Assert(startIndex >= 0);
+            Debug.Assert(endIndex >= 0);
+
+            if (!isStartIndexFromEnd && !isEndIndexFromEnd)
+            {
+                return startIndex >= endIndex
+                    ? Empty<TSource>()
+                    : source switch
+                    {
+                        IPartition<TSource> partition => partition.Skip(startIndex).Take(endIndex - startIndex),
+                        IList<TSource> list => new ListPartition<TSource>(list, startIndex, endIndex - 1),
+                        _ => new EnumerablePartition<TSource>(source, startIndex, endIndex - 1)
+                    };
+            }
+
+            return TakeIterator(source, isStartIndexFromEnd, startIndex, isEndIndexFromEnd, endIndex);
+        }
+
+        private static IEnumerable<TSource> TakeIterator<TSource>(
+            IEnumerable<TSource> source, bool isStartIndexFromEnd, int startIndex, bool isEndIndexFromEnd, int endIndex)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(startIndex >= 0);
+            Debug.Assert(endIndex >= 0);
+
+            if (isStartIndexFromEnd)
+            {
+                if (startIndex == 0)
+                {
+                    yield break;
+                }
+
+                using IEnumerator<TSource> e = source.GetEnumerator();
+                if (!e.MoveNext())
+                {
+                    yield break;
+                }
+
+                int index = 0;
+                Queue<TSource> queue = new();
+                queue.Enqueue(e.Current);
+
+                while (e.MoveNext())
+                {
+                    checked
+                    {
+                        index++;
+                    }
+
+                    if (queue.Count == startIndex)
+                    {
+                        queue.Dequeue();
+                    }
+
+                    queue.Enqueue(e.Current);
+                }
+
+                int count = checked(index + 1);
+                Debug.Assert(queue.Count == Math.Min(count, startIndex));
+
+                startIndex = count - startIndex;
+                if (startIndex < 0)
+                {
+                    startIndex = 0;
+                }
+
+                if (isEndIndexFromEnd)
+                {
+                    endIndex = count - endIndex;
+                }
+                else if (endIndex > count)
+                {
+                    endIndex = count;
+                }
+
+                Debug.Assert(endIndex - startIndex <= queue.Count);
+                for (int rangeIndex = startIndex; rangeIndex < endIndex; rangeIndex++)
+                {
+                    yield return queue.Dequeue();
+                }
```

Comment:
Does this need to be an iterator?  It seems like the `yield break;`s above could instead just `return Empty<TSource>();`, and this could instead just `return queue;`?

Reply:
Nevermind, it has to be an iterator so that it's reevaluated every time GetEnumerator is called.